File: bindings/julia/src/tensor_helpers.jl
# SPDX-License-Identifier: AGPL-3.0-or-later
# © 2025 SpiralTorch Contributors
# Part of SpiralTorch — Licensed under AGPL-3.0-or-later.

"""
    tensor_from_rows(rows)

Construct a [`Tensor`](@ref) from a vector of equally sized row vectors. Input
values are converted to `Float32` before being transferred to the runtime.
"""
function tensor_from_rows(rows::AbstractVector{<:AbstractVector})
    nrows = length(rows)
    if nrows == 0
        return Tensor(0, 0)
    end

    ncols = length(rows[1])
    for (idx, row) in enumerate(rows)
        if length(row) != ncols
            throw(ArgumentError("row $idx has length $(length(row)) but expected $ncols"))
        end
    end

    if ncols == 0
        return Tensor(nrows, 0)
    end

    buffer = Matrix{Float32}(undef, nrows, ncols)
    @inbounds for (i, row) in enumerate(rows)
        buffer[i, :] = Float32.(row)
    end
    return Tensor(buffer)
end

"""
    tensor_from_columns(columns)

Construct a [`Tensor`](@ref) from a vector of column vectors. All columns must
be the same length.
"""
function tensor_from_columns(columns::AbstractVector{<:AbstractVector})
    ncols = length(columns)
    if ncols == 0
        return Tensor(0, 0)
    end

    nrows = length(columns[1])
    for (idx, column) in enumerate(columns)
        if length(column) != nrows
            throw(ArgumentError("column $idx has length $(length(column)) but expected $nrows"))
        end
    end

    if nrows == 0
        return Tensor(0, ncols)
    end

    buffer = Matrix{Float32}(undef, nrows, ncols)
    @inbounds for (j, column) in enumerate(columns)
        buffer[:, j] = Float32.(column)
    end
    return Tensor(buffer)
end

"""
    tensor_rows(tensor)

Return a vector of row copies from the provided [`Tensor`](@ref). Each row is a
`Vector{Float32}` that can be mutated independently.
"""
function tensor_rows(tensor::Tensor)
    data = to_array(tensor)
    nrows, _ = size(data)
    rows = Vector{Vector{Float32}}(undef, nrows)
    @inbounds for i in 1:nrows
        rows[i] = Vector{Float32}(data[i, :])
    end
    return rows
end

"""
    tensor_columns(tensor)

Return a vector of column copies from the provided [`Tensor`](@ref). Each column
is a `Vector{Float32}`.
"""
function tensor_columns(tensor::Tensor)
    data = to_array(tensor)
    _, ncols = size(data)
    columns = Vector{Vector{Float32}}(undef, ncols)
    @inbounds for j in 1:ncols
        columns[j] = Vector{Float32}(data[:, j])
    end
    return columns
end

"""
    tensor_vector(tensor)

Return a flattened `Vector{Float32}` copy of the tensor contents in row-major
order.
"""
function tensor_vector(tensor::Tensor)
    array = to_array(tensor)
    return Vector{Float32}(reshape(array, :))
end

File: bindings/go/tensor_helpers.go
// SPDX-License-Identifier: AGPL-3.0-or-later
// © 2025 SpiralTorch Contributors
// Part of SpiralTorch — Licensed under AGPL-3.0-or-later.

package spiraltorch

import "fmt"

// NewTensorFromMatrix constructs a tensor from a rectangular matrix represented
// as a slice of rows. The matrix must be rectangular (all rows have the same
// length) and contain at least one row and one column. For zero-sized tensors
// prefer NewZerosTensor.
func NewTensorFromMatrix(matrix [][]float32) (*Tensor, error) {
	if matrix == nil {
		return nil, fmt.Errorf("spiraltorch: matrix cannot be nil")
	}
	rows := len(matrix)
	if rows == 0 {
		return nil, fmt.Errorf("spiraltorch: matrix must contain at least one row")
	}

	cols := len(matrix[0])
	if cols == 0 {
		return nil, fmt.Errorf("spiraltorch: matrix must contain at least one column")
	}
	data := make([]float32, rows*cols)
	offset := 0
	for i, row := range matrix {
		if len(row) != cols {
			return nil, fmt.Errorf("spiraltorch: matrix row %d has length %d, expected %d", i, len(row), cols)
		}
		copy(data[offset:offset+cols], row)
		offset += cols
	}
	return NewTensorFromDense(rows, cols, data)
}

// NewTensorFromColumns constructs a tensor from a slice of equally sized column
// vectors. The columns are interpreted in column-major order and converted to
// the row-major layout expected by SpiralTorch. The input must contain at least
// one column and each column must have at least one element. For zero-sized
// tensors prefer NewZerosTensor.
func NewTensorFromColumns(columns [][]float32) (*Tensor, error) {
	if columns == nil {
		return nil, fmt.Errorf("spiraltorch: columns cannot be nil")
	}
	cols := len(columns)
	if cols == 0 {
		return nil, fmt.Errorf("spiraltorch: columns must contain at least one column")
	}

	rows := len(columns[0])
	if rows == 0 {
		return nil, fmt.Errorf("spiraltorch: columns must contain at least one row")
	}
	for i := 1; i < cols; i++ {
		if len(columns[i]) != rows {
			return nil, fmt.Errorf("spiraltorch: column %d has length %d, expected %d", i, len(columns[i]), rows)
		}
	}

	data := make([]float32, rows*cols)
	for r := 0; r < rows; r++ {
		base := r * cols
		for c := 0; c < cols; c++ {
			data[base+c] = columns[c][r]
		}
	}
	return NewTensorFromDense(rows, cols, data)
}

// ToMatrix decodes the tensor into a dense row-major matrix. Each row in the
// returned slice owns its backing array, making it safe to mutate without
// affecting other rows.
func (t *Tensor) ToMatrix() ([][]float32, error) {
	if t == nil {
		return nil, fmt.Errorf("spiraltorch: tensor handle is nil")
	}

	rows, cols, err := t.Shape()
	if err != nil {
		return nil, err
	}
	if rows == 0 || cols == 0 {
		matrix := make([][]float32, rows)
		for r := range matrix {
			matrix[r] = make([]float32, cols)
		}
		return matrix, nil
	}

	matrix := make([][]float32, rows)
	for r := 0; r < rows; r++ {
		row := make([]float32, cols)
		if err := t.copyRowInto(r, cols, row); err != nil {
			return nil, err
		}
		matrix[r] = row
	}
	return matrix, nil
}

// Columns returns a slice of column copies from the tensor. Each column is an
// independent slice that can be mutated without affecting the underlying
// tensor or other columns.
func (t *Tensor) Columns() ([][]float32, error) {
	if t == nil {
		return nil, fmt.Errorf("spiraltorch: tensor handle is nil")
	}

	rows, cols, err := t.Shape()
	if err != nil {
		return nil, err
	}
	result := make([][]float32, cols)
	if rows == 0 || cols == 0 {
		for i := range result {
			result[i] = make([]float32, rows)
		}
		return result, nil
	}

	for c := 0; c < cols; c++ {
		column := make([]float32, rows)
		if err := t.copyColumnInto(c, rows, column); err != nil {
			return nil, err
		}
		result[c] = column
	}
	return result, nil
}

// Row returns a copy of the requested row. The returned slice owns its backing
// array and may be mutated freely.
func (t *Tensor) Row(index int) ([]float32, error) {
	if t == nil {
		return nil, fmt.Errorf("spiraltorch: tensor handle is nil")
	}

	rows, cols, err := t.Shape()
	if err != nil {
		return nil, err
	}
	if index < 0 || index >= rows {
		return nil, fmt.Errorf("spiraltorch: row index %d out of range [0,%d)", index, rows)
	}
	if cols == 0 {
		return make([]float32, 0), nil
	}

	row := make([]float32, cols)
	if err := t.copyRowInto(index, cols, row); err != nil {
		return nil, err
	}
	return row, nil
}

// Column returns a copy of the requested column. The returned slice owns its
// backing array and may be mutated freely.
func (t *Tensor) Column(index int) ([]float32, error) {
	if t == nil {
		return nil, fmt.Errorf("spiraltorch: tensor handle is nil")
	}

	rows, cols, err := t.Shape()
	if err != nil {
		return nil, err
	}
	if index < 0 || index >= cols {
		return nil, fmt.Errorf("spiraltorch: column index %d out of range [0,%d)", index, cols)
	}
	if rows == 0 {
		return make([]float32, 0), nil
	}

	column := make([]float32, rows)
	if err := t.copyColumnInto(index, rows, column); err != nil {
		return nil, err
	}
	return column, nil
}
